### PR TITLE
Update documentation on experimental mcmc module

### LINKED ISSUE #514: Provide more documentation details on experimental mcmc sampler module
Description:
-------------
Update description of the experimental mcmc module to give more context and details.

Definition of done:
-------------------
- [x] Description of experimental mcmc module in API overview is updated
- [ ] Change is reviewed and accepted

### LINKED ISSUE #515: Support burnthin in get_samples method
Description:
-------------
Allow users to remove burnin after calling `get_samples` in when using Gibbs sampling more easily.

Suggested change. Overload the dict object as "JointSamples" object, which contains multiple samples objects. Enable method `burnthin` such that users can call:
```python
sampler.get_samples().burnthin(200) # Should work in both single Samples object and JointSamples
```
The code would be this in `_samples.py`
```python
class JointSamples(dict):
    """ An object used to store samples from joint distributions. 

    This object is a simple overload of the dictionary class to allow easy access to certain methods 
    of Samples objects without having to iterate over each key in the dictionary. 
    
    """

    def burnthin(self, Nb, Nt=1):
        """ Remove burn-in and thin samples for all samples in the dictionary. Returns a copy of the samples stored in the dictionary. """
        return JointSamples({key: samples.burnthin(Nb, Nt) for key, samples in self.items()})

    def __repr__(self) -> str: 
        return "CUQIpy JointSamples Dict:\n" + \
               "-------------------------\n\n" + \
               "Keys:\n {}\n\n".format(list(self.keys())) + \
               "Ns (number of samples):\n {}\n\n".format({key: samples.Ns for key, samples in self.items()}) + \
               "Geometry:\n {}\n\n".format({key: samples.geometry for key, samples in self.items()}) + \
               "Shape:\n {}\n\n".format({key: samples.shape for key, samples in self.items()})

```
The repr would return stuff like
```bash
CUQIpy JointSamples Dict:
-------------------------

Keys:
 ['x', 's']

Ns (number of samples):
 {'x': 250, 's': 250}

Geometry:
 {'x': _DefaultGeometry1D(128,), 's': _DefaultGeometry1D(1,)}

Shape:
 {'x': (128, 250), 's': (1, 250)}
```
Definition of done:
-------------------
- [x] Suggested change is reviewed.
- [x] JointSamples object is implemented.
- [ ] Implementation is reviewed and approved.

### PR DESCRIPTION
**Pull request by nabriis**
Closes #514 
Closes #515

### CODE REVIEW

File: cuqi/experimental/mcmc/__init__.py [hunk 1]
```diff
@@ -1,4 +1,106 @@
...
+      from cuqi.distribution import DistributionGallery
+
+      # Target distribution
+      target = DistributionGallery("donut")
+
+      # Set up sampler
+      sampler = MH(target)
+
+      # Sample from the target distribution
+      samples = sampler.sample(Ns=100, Nb=100)
```
Comment (jakobsj) [inline cuqi/experimental/mcmc/__init__.py:24]: Below burnthin is used. To align with this, either use also burnthin here, or explain how the same is achived with Nb=100 (no thinning I suppose).

**Review by jakobsj — CHANGES_REQUESTED** [review]
Very nice and helpful guidance on how to use the new samplers. I added a small comment on mentioning burnthin.  I like the JointSamples very much, the usage becomes very neat.  Possibly, at a later stage, one could consider further refining the repr, for example split the printing of info for separate random variables on individual lines rather then same line in a list - I think that would look neat.

File: cuqi/experimental/mcmc/__init__.py [hunk 2]
```diff
@@ -1,4 +1,106 @@
...
+
+   There are now more options for Gibbs sampling. Previously it was only possible to sample with Gibbs for samplers `LinearRTO`, `RegularizedLinearRTO`, `Conjugate`, and `ConjugateApprox`.
+
+   Now, it is possible to define a Gibbs sampling scheme using any sampler from the `cuqi.experimental.mcmc` module.
+
+   **Example using a NUTS-within-Gibbs scheme for a 1D deconvolution problem:**
+
+   .. code-block:: python
+
+      import cuqi
```
Comment (chaozg) [inline cuqi/experimental/mcmc/__init__.py:65]: This example is great and it runs smoothly

File: cuqi/samples/_samples.py [hunk 3]
```diff
@@ -881,3 +881,23 @@ def __repr__(self) -> str:
                "Geometry:\n {}\n\n".format(self.geometry) + \
                "Shape:\n {}\n\n".format(self.shape) + \
                "Samples:\n {}\n\n".format(self.samples)
+
+class JointSamples(dict):
+    """ An object used to store samples from joint distributions.
```
Comment (chaozg) [inline cuqi/samples/_samples.py:886]:
```suggestion
    """ An object used to store samples from :class:`cuqi.distribution.JointDistribution`. 
```

File: cuqi/experimental/mcmc/__init__.py [hunk 4]
```diff
@@ -1,4 +1,106 @@
...
+"""
+Re-implementation of sampler module in a more object-oriented way.
+
+Main changes for users
+----------------------
+
+1. Sampling API
+   ^^^^^^^^^^^^
+
+   Previously one would call the `.sample` or `sample_adapt` methods of a sampler instance to sample from a target distribution and store the samples as the output as follows:
```
Comment (chaozg) [inline cuqi/experimental/mcmc/__init__.py:10]:
```suggestion
   Previously one would call the `.sample` or `sample_adapt` methods of a sampler instance at :py:mod:`cuqi.sampler` to sample from a target distribution and store the samples as the output as follows:
```

File: cuqi/experimental/mcmc/__init__.py [hunk 5]
```diff
@@ -1,4 +1,106 @@
...
+
+   There is now a flag `experimental` that can be set to `True` to use the new MCMC samplers.
+   
+   By default, the flag is set to `False` and the old samplers are used.
+
+
+3. More options for Gibbs sampling
+   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+   There are now more options for Gibbs sampling. Previously it was only possible to sample with Gibbs for samplers `LinearRTO`, `RegularizedLinearRTO`, `Conjugate`, and `ConjugateApprox`.
```
Comment (chaozg) [inline cuqi/experimental/mcmc/__init__.py:57]:
```suggestion
   There are now more options for Gibbs sampling. Previously it was only possible to sample with Gibbs for samplers  :py:class:`cuqi.sampler.LinearRTO`, :py:class:`cuqi.sampler.RegularizedLinearRTO`, :py:class:`cuqi.sampler.Conjugate`, and :py:class:`cuqi.sampler.ConjugateApprox`.
```

File: cuqi/experimental/mcmc/__init__.py [hunk 6]
```diff
@@ -1,4 +1,106 @@
...
+
+   Now, it is possible to define a Gibbs sampling scheme using any sampler from the `cuqi.experimental.mcmc` module.
```
Comment (chaozg) [inline cuqi/experimental/mcmc/__init__.py:59]:
```suggestion
   Now, it is possible to define a Gibbs sampling scheme using any sampler from the :py:mod:`cuqi.experimental.mcmc` module.
```

File: cuqi/experimental/mcmc/__init__.py [hunk 7]
```diff
@@ -1,4 +1,106 @@
...
+
+      # Sample from the target distribution
+      sampler.warmup(Nb=100)  # Explicit warmup (tuning) of sampler
+      sampler.sample(Ns=100)  # Actual sampling
+      samples = sampler.get_samples().burnthin(100) # Getting samples and removing burn-in
+
+2. Sampling API for BayesianProblem
+   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+   Bayesian problem continues to have the same API for `sample_posterior` and the `UQ` method.
```
Comment (chaozg) [inline cuqi/experimental/mcmc/__init__.py:47]:
```suggestion
   :py:class:`cuqi.problem.BayesianProblem` continues to have the same API for `sample_posterior` and the `UQ` method.
```

**Review by chaozg — CHANGES_REQUESTED** [review]
Thanks @nabriis for this nice overview of the experimental module and update on joint samples. I only left some comments on crossrefs to relevant cuqi modules/classes.

File: cuqi/experimental/mcmc/__init__.py (hunk 1 above)
Comment (nabriis) [reply cuqi/experimental/mcmc/__init__.py:24]: I updated the description to explain.

[the remark this answers is not in the document]
Comment (nabriis) [reply cuqi/experimental/mcmc/__init__.py]: Good catch. Thanks!

**Comment by nabriis** [comment]
Thanks @jakobsj and @chaozg. I believe I resolved you comments. Good catch on the sample_adapt Charlie!

**Review by jakobsj — APPROVED** [review]
Happy with the changes!

**Review by chaozg — APPROVED** [review]
LGTM

**Pull request merged**